src/Arcus.WebApi.Logging/RequestTrackingMiddleware.cs

```diff
@@ -0,0 +1,125 @@
+        /// <param name="logger">The logger to write diagnostic messages during the request tracking.</param>
+        public RequestTrackingMiddleware(
+            RequestTrackingOptions options,
+            RequestDelegate next,
+            ILogger<RequestTrackingMiddleware> logger)
+        {
+            Guard.NotNull(options, nameof(options));
+            Guard.NotNull(next, nameof(next));
+            Guard.NotNull(logger, nameof(logger));
+
+            _options = options;
+            _next = next;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Logs every incoming HTTP request.
+        /// </summary>
+        /// <param name="httpContext">The current HTTP context.</param>
+        public async Task Invoke(HttpContext httpContext)
+        {
+            var stopwatch = Stopwatch.StartNew();
+            if (_options.IncludeRequestBody)
+            {
+                httpContext.Request.EnableBuffering();
+            }
+
+            try
+            {
+                await _next(httpContext);
+            }
+            finally
+            {
+                stopwatch.Stop();
+                await TrackRequestAsync(httpContext, stopwatch.Elapsed);
+            }
+        }
+
+        private async Task TrackRequestAsync(HttpContext httpContext, TimeSpan duration)
+        {
+            try
+            {
+                IDictionary<string, StringValues> headers = 
+                    _options.IncludeRequestHeaders
+                        ? SanitizeRequestHeaders(httpContext.Request.Headers) ?? new Dictionary<string, StringValues>()
+                        : new Dictionary<string, StringValues>();
+                
+                
+                IDictionary<string, StringValues> body = 
+                    _options.IncludeRequestBody
+                        ? await SanitizeRequestBodyAsync(httpContext.Request.Body) ?? new Dictionary<string, StringValues>()
+                        : new Dictionary<string, StringValues>();
+                
+                Dictionary<string, object> logContext = headers.Concat(body).ToDictionary(kv => kv.Key, kv => (object) kv.Value);
+                _logger.LogRequest(httpContext.Request, httpContext.Response, duration, logContext);
+            }
+            catch (Exception exception)
+            {
+                _logger.LogCritical(exception, "Failed to track request");
+            }
+        }
+
+        /// <summary>
+        /// Extracts information from the given HTTP <paramref name="requestHeaders"/> to include in the request tracking context.
+        /// </summary>
+        /// <param name="requestHeaders">The headers of the current HTTP request.</param>
+        protected virtual IDictionary<string, StringValues> SanitizeRequestHeaders(IDictionary<string, StringValues> requestHeaders)
+        {
+            return requestHeaders.Where(header => !_options.OmittedHeaderNames.Contains(header.Key));
```

Comment:
But this is just personal preference, you can ignore if you want
```suggestion
            return requestHeaders.Where(header => _options.OmittedHeaderNames.Contains(header.Key) == false);
```

Reply:
Hmm, 🤔 , flipping booleans is less work than a new condition. But I see that people would like this more. I'll change it for you. But again, something I would normally not do.

Reply:
`!` just sucks from a readabilty perspective - You need to clearly look for it to notice it in my opinion; but maybe I'm reviewing too much lately 😂